net/l2util: add eui_short/48/64_get()

benpicco opened the pull request:
### Contribution description

Based on the discussion in #12592 the consensus seemed that there should be an additional layer on top of luid to dispatch L2 addresses.

Maybe it could be as simple as that?

If your board provides a EUI64, with this all you have to do is implement
```C
size_t board_get_eui64(netdev_t *netdev, eui64_t *addr) {
    …
}
```

and then figure out a way to dispatch the address to the right `netdev`.
However, for the most common case where there is only one interface, you can simple ignore `netdev`.

### Testing procedure

Nothing to test yet.

### Issues/PRs references
depends on #12592
addresses #12616

maribu commented on sys/net/link_layer/l2util/l2util.c:
```diff
@@ -238,5 +239,52 @@ int l2util_ndp_addr_len_from_l2ao(int dev_type,
+    return 0;
+}
+
+void l2util_get_eui_short(netdev_t *netdev, uint16_t *addr)
```
Couldn't we mark this function as `weak` and let it unconditionally call `luid_get_short()`? I think that could result in smaller ROM. Or am I underestimating the optimisation of the compiler?

benpicco replied:
The idea is to fall back to `luid_get_short()` if the real addresses are exhausted / there is none available for `netdev`.

benpicco replied:
I could make those functions `static inline` though, could help a bit.

maribu replied:
With LTO I expect the same result either way. But I'm personally afraid to use it, as calls to external functions are memory barriers without LTO, and stop being so with. I opened a few PRs to add explicit memory barriers for `irq_disable()` and `irq_restore()`, which should be wrapped arround the majority of cases where barriers are needed anyway. I think the are now all merged and they had no effect on the generated code without LTO. But there are likely still some bugs present that only manifest wih LTO.

miri64 commented on sys/net/link_layer/l2util/l2util.c:
```diff
@@ -238,5 +239,52 @@ int l2util_ndp_addr_len_from_l2ao(int dev_type,
     return -ENOTSUP;
 }
 
+size_t __attribute__((weak)) board_get_eui_short(netdev_t *netdev, uint16_t *addr)
```
How does the implementation know which `netdev` is which?

benpicco replied:
Excellent question!
One way would be extending the driver `params_t` by an identifier, or check which SPI device is used.
Currently there is no general way to do this in RIOT.

But I deemed this the most future proof solution.

maribu replied:
The SPI bus will only help if at most one network device is connected to each SPI bus. On boards without DMA support for SPI buses, there is no advantage in putting each device on a different bus, as the MCU still can only access one at a time. And some boards only have one SPI bus available anyway.

benpicco replied:
There is also `cs_pin` or some other driver specific pin, that's all board / driver specific.

maribu replied:
OK. A stupidly simple idea: Pass the name of the driver and the index in the `<foo>_params_t` array. As the user will have supply the `<foo>_params_t` array manually, she/he has to be aware of the order in there.

As mitigation of the overhead in case this info is not needed, this function could be `static inline` in case of default (LUID based) generation, which ignores the name and index. The compiler should be smart enough to optimize out both uneeded arguments.

The hook would need to be implemented as a pseudomodule, let's say `l2addr_custom`. The code could look something like this:

```C
extern int custom_generate_eui64(eui64_t *addr, const char *driver, unsigned params_idx)

static inline void l2util_generate_eui64(eui64_t *addr, const char *driver, unsigned params_idx)
{
    if (IS_ACTIVE(MODULE_L2ADDR_CUSTOM)) {
        if (!custom_generate_eui64(addr, driver, params_idx)) {
            /* custom L2 addr successfully provided, we're done */
            return;
        }
        /* custom L2 addr not provided, fall back to LUID */
    }
    luid_get_eui64(addr);
}
```

benpicco replied:
I like the idea. A pseudomodule feels a bit clunky, why not just have
`#define BOARD_PROVIDES_EUI64    (1)` in `board.h`? That's where the `<foo>_params_t` array lives too, so it's all in one place.

maribu replied:
I see a user with a single 802.15.4 transceiver and a magic globally unique EUI-64 chip as the default use case for this. So the driver of the magic EUI-64 chip could just add that module as its dependency and provide a `custom_generate_eui64()` as weak symbol, that sets its EUI-64 on its first call and returns e.g. `-EADDRINUSE` on subsequent calls.

This would allow users with more transceivers to provide a custom `custom_generate_eui64()`(*) as strong symbol. And in the default use case of a single 802.15.4 transceiver, the user would only need to add the driver of the magic EUI-64 chip to `USEMODULE` and would be done.

(*) I heard you like customization, so I added a customize hook in the customize hook, so that you can customize while customize.

benpicco replied:
I see - so you just select your radio and your eui64-chip driver and as long as there is no ambiguity, no board support code is needed - neat!

benpicco replied:
That "eui64-chip driver" would actually be an [EEPROM](http://ww1.microchip.com/downloads/en/devicedoc/Atmel-8807-SEEPROM-AT24MAC402-602-Datasheet.pdf) [driver](https://github.com/RIOT-OS/RIOT/pull/11929).

Now I'm wondering if it would make more sense to add a function to the EEPROM API (#12091) to get a MAC address from an EEPROM device or if the board should just call the function of the specific EEPROM driver directly.

miri64 replied:
> Now I'm wondering if it would make more sense to add a function to the EEPROM API (#12091) to get a MAC address from an EEPROM device or if the board should just call the function of the specific EEPROM driver directly.

Not 100% sure what you are trying to say, but adding yet another function stubb to this *abstraction* layer, seems counter-intuitive to me.

maribu replied:
> Now I'm wondering if it would make more sense to add a function to the EEPROM API (#12091) to get a MAC address from an EEPROM device or if the board should just call the function of the specific EEPROM driver directly.

I'd say adding a separate API for retrieving factory provided globally unique EUI-48/EUI-64 would be a more general approach. E.g. the ESP32 has an EUI-48 integrated (I think I so this being used as the CPU-ID for that platform), but no EEPROM. To me, there would be no reason to not allow a driver for an EEPROM chip that additionally has an EUI integrated to not just implement two APIs.

benpicco replied:
There is a problem with the `params_idx` idea:
The function would be called in `void at86rf2xx_reset(at86rf2xx_t *dev)`. `dev->param` exists, but it's not a pointer to the `at86rf2xx_params_t` array in ROM but rather a copy of it in RAM, so there is no way to calculate the index. (Except for `memcmp()`, but if the content of the struct doesn't change, why copy it to RAM in the first place?)

maribu replied:
The is an design flaw in the driver that I tackeled in my open PR. That function should not be called in reset but in init. A copy of the address is hold in RAM anyway, which can be used to restore the devices address during reset.

benpicco replied:
Alright, I'll update the API then.

benpicco commented:
Rebased & renamed the functions to better fit with the other functions in `eui48.h`/`eui64.h`

miri64 commented on sys/include/net/eui48.h:
```diff
@@ -131,6 +131,35 @@ static inline void eui48_clear_group(eui48_t *addr)
+/**
+ * @brief Board-specific function to supply an EUI-48 to a netdev
+ *
+ *        Implement this function in your board code if the board
```
```suggestion
 * @note  Implement this function in your board code if the board
```
?

miri64 commented on sys/include/net/eui64.h:
```diff
@@ -83,6 +83,66 @@ static inline void eui64_clear_group(eui64_t *addr)
+/**
+ * @brief Board-specific function to supply a short address to a netdev
+ *
+ *        Implement this function in your board code if the board
```
```suggestion
 * @note  Implement this function in your board code if the board
```

benpicco replied:
Added the `@note` and a `@warning` not to call the function directly.

miri64 commented on sys/include/net/eui48.h:
```diff
@@ -131,6 +131,42 @@ static inline void eui48_clear_group(eui48_t *addr)
+ * @param[out] addr     The generated EUI-48 address
+ *
+ */
+static inline void eui48_get(netdev_t *netdev, eui48_t *addr)
```
The names are somewhat misleading IMHO. With `euiXX_get` I would expect that the functions return the `euiXX` of the device (so one that is already preconfigured). However, they generate the `euiXX` for the device (to configure them at the device). How about `eui48_generate()`?

miri64 commented on sys/include/net/eui48.h:
```diff
@@ -131,6 +131,42 @@ static inline void eui48_clear_group(eui48_t *addr)
+ * @param[out] addr     The dedicated address for the netdev
+ *
+ */
+size_t board_get_eui48(netdev_t *netdev, eui48_t *addr);
```
`eui48_get_from_board()`?

miri64 commented on sys/include/net/eui64.h:
```diff
@@ -83,6 +83,80 @@ static inline void eui64_clear_group(eui64_t *addr)
+ * @param[out] addr     The generated short address
+ *
+ */
+static inline void eui_short_get(netdev_t *netdev, uint16_t *addr)
```
Why is this in `eui64.h`. Neither is the short address of IEEE 802.15.4 an EUI, nor has it anything to do with an EUI-64 (except that they are both used in IEEE 802.15.4 as addresses). This function and `board_get_eui_short` don't belong here. You've put the implementation for `board_get_…()` already in `l2util` (which given that `l2util` might not be compiled in in any case is somewhat of a bargain IMHO, but let's discuss this elsewhere), so why not put it there as `l2util_generate_short_addr()`?

benpicco replied:
Then it should also be `l2util_generate_euiXX()` for consistency.

miri64 replied:
Yes, i wrote the other comments before I realized, that the other functions are also misplaced.

benpicco commented:
I moved them back to `l2util` once again.

miri64 commented on sys/include/net/l2util.h:
```diff
@@ -133,6 +133,116 @@ int l2util_ipv6_iid_to_addr(int dev_type, const eui64_t *iid, uint8_t *addr);
+ * @param[out] addr     The dedicated address for the netdev
+ *
+ */
+size_t short_addr_get_from_board(netdev_t *netdev, uint16_t *addr);
```
Name-spacing? How about

```suggestion
size_t l2util_short_addr_from_board(netdev_t *netdev, uint16_t *addr);
```

Also for the other functions.

benpicco replied:
Yea since the actual implementation will be in the respective folder for the individual boards, I thought the `board_*` namespace would be more appropriate.

miri64 replied:
But it doesn't have this either ;-)

benpicco replied:
Changed it back to `board_get_short_addr()`

maribu commented:
Let's define what we expect as use cases by likelyhood. I'd say there are three scenarios:

1. There are no EUIs provided at all, LUID has to be used
2. There is a single EUI chip on the board
3. There are multiple EUI chips on the board

I believe that this is also the order of likelihood. I think that this PR handles case 1. and 3. very well, but case 2. could be handled better: If only a single EUI driver is loaded, this could provide the custom EUI function as weak symbol, which returns the provided EUI on the first call and an error on all subsequent. This would require no board support (except for adding e.g. `USEMODUE += at24mac`). If multiple EUI chips are present, the board would need to provide the function as strong symbol that matches the driver name and index to the correct EUI provider.

benpicco commented:
As for 2., I could implement

```C
size_t board_get_eui48(const char *driver, unsigned idx, eui48_t *addr) {
    (void) driver;
    /* Maybe at24mac_get_euiXX() should just return the size */
    if (at24mac_get_eui48(idx, addr)) {
        return 0;
    }
    return sizeof(*addr);
}
```
in #12746.

benpicco commented:
I think we should just drop that 'automatic' MAC chip <-> transceiver pairing and leave it to the board to implement `board_get_euiXX()`.

If at some point we come up with a good solution we can still convert it to that, but let's start with the simple implementation first.

miri64 commented on sys/include/net/l2util.h:
```diff
@@ -133,6 +134,125 @@ int l2util_ipv6_iid_to_addr(int dev_type, const eui64_t *iid, uint8_t *addr);
+ *
+ * @warning Don't call this function directly, use @ref l2util_generate_short_addr() instead.
+ *
+ * @param[in] driver    The driver name of the netdev
```
Please relight my memory: What is the driver name of the netdev?

benpicco replied:
The idea was that a EUI would be assigned to a certain device/driver.
E.g. both `samr21-xpro` and `same54-xpro` contain a EUI-64 resp. EUI-48.
On `samr21-xpro` that should only be used for the first device of the `at86rf2xx` driver…

It's not pretty but so far we haven't come up with a better solution to assign IDs to netdevs.

miri64 replied:
> It's not pretty but so far we haven't come up with a better solution to assign IDs to netdevs.

Mhhhh its not just not pretty, it can be very costly. Both the string will cost memory, and string operations aren't the most efficient either. :-/

miri64 replied:
Would it make sense to use [`NETDEV_TYPE_`](https://doc.riot-os.org/group__drivers__netdev__api.html) here?

miri64 replied:
If the `netdev` allocations would be in an [xfa](https://github.com/RIOT-OS/RIOT/pull/9105) this would be simpler :-/.

benpicco replied:
> Would it make sense to use `NETDEV_TYPE_` here?

Not really - the idea was that if you e.g. attach a mrf24j40 to a `samr21-xpro`, thet at86rf2xx radio would still get the MAC address that's printed on the board.

A better option would be if netdevs had IDs, but right now they just have a PID and that changes depending on what modules are used.

miri64 replied:
> A better option would be if netdevs had IDs, but right now they just have a PID and that changes depending on what modules are used.

Where do they have a PID? `gnrc_netif`s have a PID, but those are not `netdev`s ;-).

benpicco replied:
Oh yea, got confused there - well then afaik there is no way to uniquely identify a `netdev` ;-)

miri64 replied:
> [1] I'd be fine with just handing out UIDs in a first-come first-serve basis.
That's the common case anyway and it would make things much simpler.

Mh, yeah, why don't we ask of a `netdev` driver to provide some globally unique numeric value (e.g. `AT86RF2XX_DEVUID`) in their header. If a board then implements `board_get_…`, it can use that value to identify the devices.

benpicco replied:
Then if you have more than one `netdev`s with the same driver it's still first-come first-serve (as with the string based approach)

miri64 replied:
What do you mean by first-come first serve? I thought `idx` is in line with the `params` array (which given that at the moment `auto_init` calls the setup functions in that order, should be the case).

benpicco replied:
> even if locally in the driver we don't know the position

I can think of three options:

 - `at86rf2xx_setup(at86rf2xx_t *dev, const at86rf2xx_params_t *params)` -> `at86rf2xx_setup(at86rf2xx_t *dev, uint8_t idx)` - a bit tedious but probably the cleanest solution
- `(params - &at86rf2xx_params) / sizeof(*params)` to get the index (urgh)
- introduce a local counter in the driver (and waste a precious byte of RAM)

miri64 replied:
Why not provide a board-unique ID with `params` btw?

miri64 replied:
For this we would have the migration benefit, that this could be done driver by driver.

benpicco replied:
> Why not provide a board-unique ID with `params` btw?

- we have to touch every board
- we waste RAM and ROM for information that is already implicitly available

miri64 replied:
>     * we waste RAM and ROM for information that is already implicitly available

I thought it is not. If it is why use the (more RAM heavy) name?



>     * we have to touch every board

If we don't use the current approach, I believe we don't have. Also: only every board that provides a netdev `params`.

benpicco replied:
> I thought it is not. If it is why use the (more RAM heavy) name?

Well we just have to touch every netdev driver (to do the index -> params mapping inside the driver instead of auto_init). I just want to be sure that's the right way to go before I shave that Yak 😉

miri64 replied:
> I just want to be sure that's the right way to go before I shave that Yak wink

See https://github.com/RIOT-OS/RIOT/pull/12641#issuecomment-580240006. As far as I understand `board.c` might not exist anymore in the near future :-/.

miri64 commented:
Apparently, there is some rework to `board.c` waiting to be proposed. @kaspar030 could you maybe have a look, at the core idea of this PR is still relevant with your rework in mind.

benpicco commented:
I've converted `at86rf2xx` as an example and implemented `board_get_eui64()` for `samr21-xpro`.
The HWaddr should now match the one printed backside of the board.

miri64 commented:
@benpicco can you give this a rebase and status update on the WIP state of this PR?

benpicco commented:
Requires #13746 for a *proper* solution, but I don't have time to shave that yak now.
If we want to solve the common 'board provides EUI, interface should use EUI' case and we don't care about the order for the multi-EUI / multi-Interface case, we can just implement `board_get_eui…()` as initially proposed here.

benpicco commented:
superseded by #13746
